GitHub Actions를 활용한 CD 파이프라인 작성

hye-on opened the pull request:
### 📝 작업 내용

cd 작업했습니다.

## 💬 리뷰 요구사항(선택)

전체적인 플로우 봐주세요!
더 좋은 문제 해결법이 있을 것 같은데,, 혹시 아는 거 있으면 알려주세요! :)


### ⛔ serverless 버전4에서는 로그인 필수
`Error: You must sign in or use a license key with Serverless Framework V.4 and later versions. Please use "serverless login".`
=>Serverless Framework v4부터 로그인이나 라이선스 키가 필요

[실패한 github action](https://github.com/Mash-Up-Node/mashup-nest-mercurius-fastify/actions/runs/14576286652/job/40883072528)

=> 🔑  3.38.0로 버전 낮춰서 해결
-> 나중에 라이선스 키 받으면 4로 올려도 좋을 것 같아요

### ⛔ 플러그인 충돌
`TypeError: Cannot redefine property: _serverlessExternalPluginName`
[실패한 github action](https://github.com/Mash-Up-Node/mashup-nest-mercurius-fastify/actions/runs/14577259684/job/40886069751)

**시도한 방법**
https://forum.serverless.com/t/cannot-redefine-property-serverlessexternalpluginname/20666
https://github.com/serverless/serverless/issues/12936
-> 이거 보고 노드 버전 수정해봤는데 해결안되었습니다. 

=> 🔑 
serverless.yml에서 플러그인 주석 처리해서 해결
CI/CD 배포 단계에서 필요없는 플러그인 인것같아서 주석처리 했는데 문제 없으면 주석 삭제해서 커밋하겠습니다.
```
plugins:
#  - serverless-offline
#  - serverless-dotenv-plugin
```

### ⛔ 패키지 사이즈 너무 커서 람다 배포안되는 문제
[실패한 github action](https://github.com/Mash-Up-Node/mashup-nest-mercurius-fastify/actions/runs/14587606873/job/40915858895)

=> 🔑  `serverless.yml`에 - '!node_modules/.pnpm/**' 추가
임시방편 인 것 같아서 조금 더 이유를 찾아볼께요. 

### ⛔ 서울 리전 지원안되는 문제

https://www.serverless.com/framework/docs/guides/dashboard/cicd/running-in-your-own-cicd#configure-environment-variables

<img width="500" alt="image" src="https://github.com/user-attachments/assets/6c0f056c-cc2f-4f04-b1f2-29db48804952" />


SERVERLESS_ACCESS_KEY로 하면 대시보드에 엑세스하는데 대시보드 모드가 서울리전이 지원안되는 것 같아요.

[실패한 github action](https://github.com/Mash-Up-Node/mashup-nest-mercurius-fastify/actions/runs/14587141417/job/40914651046)

=> 🔑 `SERVERLESS_ACCESS_KEY` 사용안하고 AWS 자격 증명(AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY)만 사용

아래 코드 주석처리 해서 대시보드 기능 비활성화
(혹시 아니면 말해주세욤)

`serverless.yml`
```
#org: hospace
#app: mashup-node
```
## 📚 참고할만한 자료(선택)

Commit:
feat: cd 스크립트 작성

Commit:
chore: cd스크립트  workflows폴더로 이동

Commit:
chore: 버전 수정

Commit:
chore: serverless 버전 3.38.0으로 고정 및 credential 과정 삭제

- serverless v4부터 로그인/라이선스 키가 필요해 v3.38.0으로 고정
- AWS 인증은 환경 변수만으로 처리하도록 변경

Commit:
chore: prisma deploy과정에 Prisma CLI 설치 추가

Commit:
chore: 프리즈마 스키마 파일 위치 설정

Commit:
chore: 경로 추가

Commit:
chore: 환경 변수 이름 수정

Commit:
fix: serverless 배포 명령어 수정. 플러그인 충돌 해결

- npx --no-install 옵션을 사용하여 serverless 버전 충돌 문제 해결

Commit:
chore: 플러그인 주석

Commit:
chore: 서버리스 버전 수정

Commit:
chore: 노드 버전 수정

hye-on committed:
chore: 플러그인 충돌로 인해 주석처리

hye-on committed:
chore: AWS 자격 증명 설정

hye-on committed:
chore: serverless secret key 추가

hye-on committed:
chore: 서버리스 버전 수정

hye-on committed:
chore: 대시보드 비활성화

hye-on committed:
chore: 대시보드 모드 비활성화

hye-on committed:
chore: --no-console 제거

hye-on committed:
chore: pakage할때 제외 pnpm제외

Ho-s commented on .github/deploy.yml:
```diff
@@ -0,0 +1,82 @@
+        run: yarn build
+        
+      - name: Upload build artifact
+        uses: actions/upload-artifact@v3
```
이 친구는 뭐 할 때 사용하는 액션인가요?

hye-on replied:
빌드 결과물을 저장하고 공유하는 사용합니다!
다른 워크플로우 작업이나 단계에서 빌드 결과물을 공유하기 위해 사용했습니다!

현재 2개 두 개의 별도 작업으로 구성되어 있습니다.
- validate 작업: 코드를 체크아웃하고, 의존성을 설치하고, 빌드하는 작업
- deploy 작업: 빌드된 결과물을 AWS Lambda에 배포하는 작업

Ho-s commented on .github/deploy.yml:
```diff
@@ -0,0 +1,82 @@
+          AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          DB_URL: ${{ secrets.DB_URL }}
+          DB_HOST: ${{ secrets.DB_HOST }}
```
기존에는 `PORT`와 `DB_URL` 만 쓰고 있습니다!

hye-on replied:
[d90b977](https://github.com/Mash-Up-Node/mashup-nest-mercurius-fastify/pull/10/commits/d90b977913548c1f65ec945465b23b27caa7e3a5) 제거했습니다!

Ho-s commented on .github/workflows/deploy.yml:
```diff
@@ -56,11 +56,8 @@ jobs:
           
       - name: Install Serverless Framework
-        run: npm install -g serverless
+        run: npm install -g serverless@3.38.0
```
local serverless도 버전 맞춰주면 좋을 것 같습니다!

Ho-s commented on serverless.yml:
```diff
@@ -5,7 +5,7 @@ service: nest-graphql-mercurius-fastify
 plugins:
   - serverless-offline
-  - serverless-dotenv-plugin
+ # - serverless-dotenv-plugin
```
local에서 사용해야 하니,

github ci 중에 `echo >>` 를 활용하여 .env를 만들어주거나
serverless.yml이 아닌 serverless.js 로 분기처리를 하는 방법들이 있을 거 같습니다!

혹시, 주석 처리한 다른 이유가 있으신가요?

Ho-s commented on serverless.yml:
```diff
@@ -4,7 +4,7 @@ app: mashup-node
 service: nest-graphql-mercurius-fastify
 
 plugins:
-  - serverless-offline
```
offline은 현재는 필요 없어 다른 브랜치에서 삭제한 상태입니다!

Ho-s commented on serverless.yml:
```diff
@@ -1,5 +1,5 @@
-org: hospace
-app: mashup-node
+#org: hospace
```
🔥

Ho-s commented on serverless.yml:
```diff
@@ -27,6 +27,7 @@ package:
     - '!**'
     - 'dist/**'
     - 'node_modules/**'
+    - '!node_modules/.pnpm/**'
```
pnpm이 들어간 이유가 궁금합니다!

Ho-s reviewed (commented):
빠르고 정확한 작업... 고군 분투의 흔적이 보입니다! 너무 고생하셨어요! 🔥🔥🔥🔥🔥

labyrinth30 commented on .github/workflows/deploy.yml:
```diff
@@ -0,0 +1,85 @@
+          name: build-artifact
+          path: |
+            dist/
+            node_modules/
```
코파일럿도 이 이야기 같은데 빌드된 걸 올리는 과정에서 node_modules까지 올려도 되는걸까요?

labyrinth30 reviewed (commented):
서버리스로 CD하는 건 저도 익숙치 않네요. 진행하시면서 엄청 고생하신 것 같습니다!
토요일에 같이 이야기하면 좋을 것 같습니다!

Ho-s committed:
chore: resolve conflicts

Ho-s committed:
chore: remove @generated from code config

Ho-s committed:
fix: build type error

Ho-s committed:
feat: bcrypt -> bcryptjs로 변경

- lambda 환경에서 bcrypt는 호환 이슈가 있어요
ref: https://github.com/kelektiv/node.bcrypt.js/issues/505#issuecomment-353135828

Ho-s committed:
feat: 필요한 환경변수 재설정

Ho-s committed:
feat: aws-lambda 를 serverless 배포에서 제외

Ho-s committed:
chore: only import type

labyrinth30 commented on package.json:
```diff
@@ -36,8 +36,7 @@
     "@prisma/client": "^6.5.0",
-    "aws-lambda": "^1.0.7",
-    "bcrypt": "^5.1.1",
+    "bcryptjs": "^3.0.2",
```
ref 확인했습니다

labyrinth30 commented on src/common/decorators/current-user.decorator.ts:
```diff
@@ -1,9 +1,12 @@
-    return ctx.getContext().req.user;
+  (_: unknown, context: ExecutionContext) => {
+    const gqlContext = GqlExecutionContext.create(context);
+    const { user } = gqlContext.getContext<GraphQLContext>();
```
제네릭으로 타입 명시하니까 타입 안정성도 높이고, 명확성과 유지보수성도 높아지는 것 같습니다👍

labyrinth30 commented on src/auth/auth.service.ts:
```diff
@@ -1,7 +1,7 @@
 import { ConfigService } from '@nestjs/config';
 import { PrismaService } from '../prisma/prisma.service';
-import * as bcrypt from 'bcrypt';
+import * as bcrypt from 'bcryptjs';
```
ref 이슈 확인했습니다.

Ho-s committed:
feat: 민감한 정보 masking

Ho-s committed:
feat: use dotenv plugin

Ho-s committed:
feat: build / deploy workflow 나눔

- npx prisma generate 는 삭제 -> yarn install 시 prisma 내부 postinstall 로 shcema에 따라 자동 생성

Ho-s committed:
feat: install serverless in local

Ho-s committed:
chore: remove no use

hye-on committed:
chore: 필요없는 환경변수 제거

hye-on committed:
chore: 주석 삭제

hye-on commented:
@Ho-s 수정 감사합니다 ! 👍

labyrinth30 commented on .github/workflows/deploy.yml:
```diff
@@ -0,0 +1,43 @@
+        run: yarn install --frozen-lockfile
+
+      - name: Run Prisma migrations
+        run: npx prisma migrate deploy 2>&1 | grep -v 'Datasource "db":' | grep -v 'PostgreSQL database'
```
덕분에 쉘 명령어 다시 공부했네요ㅋㅋㅋㅋ 불필요한 정보 마스킹하는 것 잘 배워갑니다!

Ho-s replied:
고마워 지피티야!

labyrinth30 commented on serverless.yml:
```diff
@@ -30,3 +27,4 @@ package:
     - 'node_modules/.prisma/client/libquery_engine-rhel-*'
     - '!node_modules/prisma/libquery_engine-*'
     - '!node_modules/@prisma/engines/**'
+    - '!node_modules/aws-lambda'
```
👍

labyrinth30 commented on .github/workflows/validate.yml:
```diff
@@ -0,0 +1,29 @@
+name: Run build & test
```
👍

labyrinth30 commented on .github/workflows/deploy.yml:
```diff
@@ -0,0 +1,43 @@
+name: Deploy serverless application
```
cd 과정을 여러 개로 나누는 걸 처음 봤는데, 유연하게 사용할 수 있는 게 큰 장점인 것 같아요 👍

Ho-s replied:
장단이 있는 것 같습니다.

책임 분리 측면에서는 장점을 가집니다.
기존에 @hye-on 님이 해주신 방식의 경우에는 중복 로직을 제거할 수 있다는 장점을 가졌었죠!

상황에 따라 선택하면 좋아보입니다!

labyrinth30 reviewed (changes requested):
병호형이 말한 것처럼 @generated만 없어지면 좋을 것 같습니다!

Ho-s committed:
chore: remove this branch from workflow triggers

Ho-s committed:
chore: remove useless

labyrinth30 commented:
🚀

Ho-s reviewed (approved):
🔥🔥🔥🔥🔥
